File: jsh/lib/ws/ws.js
'use strict';

const process = require('/lib/process');
const EventEmitter = require('/lib/events');
const _ws = require('@jsh/ws');

// events: "open", "close", "message", "error"
class WebSocket extends EventEmitter {
    constructor(url, protocols) {
        super();
        if (url === undefined || typeof url !== 'string') {
            throw new TypeError('URL must be a string, got ' + typeof url);
        }
        const normalizedProtocols = normalizeProtocols(protocols);
        this.raw = _ws.NewWebSocket(this, url, normalizedProtocols, process.dispatchEvent);
        this.readyState = WebSocket.CONNECTING;
        this.url = url;
        this.protocol = '';

        setImmediate(() => {
            try {
                const err = this.raw.connect();
                if (err instanceof Error) {
                    this.emit('error', err);
                } else {
                    this.protocol = typeof this.raw.protocol === 'function' ? this.raw.protocol() : '';
                    this.readyState = WebSocket.OPEN;
                    this.emit('open');
                }
            } catch (e) {
                this.emit('error', new Error(e.value || e.message || String(e)));
            }
        });
    }

    static _fromAccepted(raw, url = '') {
        const socket = Object.create(WebSocket.prototype);
        socket._events = {};
        socket._maxListeners = 10;
        socket.raw = raw;
        socket.readyState = WebSocket.OPEN;
        socket.url = url;
        socket.protocol = typeof socket.raw.protocol === 'function' ? socket.raw.protocol() : '';
        if (typeof socket.raw.onMessage !== 'function' || typeof socket.raw.onClose !== 'function' || typeof socket.raw.onError !== 'function') {
            throw new TypeError('accepted websocket native hooks are not available');
        }
        socket.raw.onMessage((event) => {
            socket.emit('message', event);
        });
        socket.raw.onError((error) => {
            socket.emit('error', error);
        });
        socket.raw.onClose((event) => {
            socket.readyState = WebSocket.CLOSED;
            socket.emit('close', event);
        });
        return socket;
    }

    send(data) {
        try {
            this.raw.send(WebSocket.TextMessage, data);
        } catch (err) {
            this.emit('error', err);
        }
    }
    close() {
        if (this.readyState === this.CLOSED) {
            return;
        }
        this.readyState = this.CLOSING;
        this.raw.close();
        this.readyState = this.CLOSED;
        this.emit('close');
    }

    static CONNECTING = 0;
    static OPEN = 1;
    static CLOSING = 2;
    static CLOSED = 3;

    static TextMessage = 1;
    static BinaryMessage = 2;
}

function normalizeProtocols(protocols) {
    if (protocols === undefined || protocols === null) {
        return [];
    }
    if (typeof protocols === 'string') {
        return [protocols];
    }
    if (Array.isArray(protocols)) {
        return protocols.filter((value) => typeof value === 'string');
    }
    throw new TypeError('Protocols must be a string or an array of strings');
}

function normalizeHeaders(headers) {
    const normalized = {};
    if (!headers || typeof headers !== 'object') {
        return normalized;
    }
    for (const key in headers) {
        normalized[String(key).toLowerCase()] = headers[key];
    }
    return normalized;
}

function normalizeRequest(rawReq) {
    if (!rawReq) {
        return rawReq;
    }
    const readProperty = (camelName, pascalName, fallback) => {
        if (rawReq[camelName] !== undefined) {
            return rawReq[camelName];
        }
        if (rawReq[pascalName] !== undefined) {
            return rawReq[pascalName];
        }
        return fallback;
    };
    const queryMethod = typeof rawReq.query === 'function'
        ? rawReq.query.bind(rawReq)
        : (typeof rawReq.Query === 'function' ? rawReq.Query.bind(rawReq) : null);
    const getHeaderMethod = typeof rawReq.getHeader === 'function'
        ? rawReq.getHeader.bind(rawReq)
        : (typeof rawReq.GetHeader === 'function' ? rawReq.GetHeader.bind(rawReq) : null);
    const hasHeaderMethod = typeof rawReq.hasHeader === 'function'
        ? rawReq.hasHeader.bind(rawReq)
        : (typeof rawReq.HasHeader === 'function' ? rawReq.HasHeader.bind(rawReq) : null);
    const headers = normalizeHeaders(readProperty('headers', 'Headers', {}));
    const rawUrl = readProperty('url', 'URL', '');
    const request = {
        url: rawUrl,
        method: readProperty('method', 'Method', 'GET'),
        headers,
        rawHeaders: readProperty('rawHeaders', 'RawHeaders', []),
        path: readProperty('path', 'Path', ''),
        host: readProperty('host', 'Host', headers.host || ''),
        requestUri: readProperty('requestURI', 'RequestURI', rawUrl),
        httpVersion: String(readProperty('proto', 'Proto', 'HTTP/1.1')).replace(/^HTTP\//, ''),
        complete: true,
        remoteAddress: readProperty('remoteAddress', 'RemoteAddress', ''),
        socket: {
            remoteAddress: readProperty('remoteAddress', 'RemoteAddress', ''),
        },
        getHeader(name) {
            if (getHeaderMethod) {
                return getHeaderMethod(name);
            }
            return headers[String(name).toLowerCase()];
        },
        hasHeader(name) {
            if (hasHeaderMethod) {
                return hasHeaderMethod(name);
            }
            return String(name).toLowerCase() in headers;
        },
        query(name) {
            return queryMethod ? queryMethod(name) : undefined;
        },
    };
    return {
        ...request,
    };
}

function createVerifyClient(options) {
    if (typeof options.verifyClient !== 'function') {
        return () => true;
    }
    return (rawReq) => {
        const req = normalizeRequest(rawReq);
        const result = options.verifyClient({
            origin: req.getHeader('origin'),
            req,
        });
        return result !== false;
    };
}

function createHandleProtocols(options) {
    if (typeof options.handleProtocols !== 'function') {
        return () => '';
    }
    return (protocols, rawReq) => {
        const req = normalizeRequest(rawReq);
        const selected = options.handleProtocols(protocols, req);
        return typeof selected === 'string' ? selected : '';
    };
}

class WebSocketServer extends EventEmitter {
    constructor(options = {}) {
        super();
        if (!options || typeof options !== 'object') {
            throw new TypeError('WebSocketServer options must be an object');
        }
        if (!options.server || typeof options.server._attachWebSocket !== 'function') {
            throw new TypeError('WebSocketServer requires an http.Server instance');
        }

        this.options = options;
        this.server = options.server;
        this.path = options.path || '/';
        this.clients = options.clientTracking === false ? null : new Set();
        this.closed = false;
        this._verifyClient = createVerifyClient(options);
        this._handleProtocols = createHandleProtocols(options);

        this.server._attachWebSocket(this.path, this._verifyClient, this._handleProtocols, (rawSocket, rawReq) => {
            try {
                if (this.closed) {
                    rawSocket.close();
                    return;
                }

                const request = normalizeRequest(rawReq);
                const socket = WebSocket._fromAccepted(rawSocket, request && request.url ? request.url : '');

                if (this.clients) {
                    this.clients.add(socket);
                    const cleanup = () => {
                        this.clients.delete(socket);
                        socket.removeListener('close', cleanup);
                    };
                    socket.on('close', cleanup);
                }

                this.emit('connection', socket, request);
                socket.emit('open');
            } catch (err) {
                this.emit('error', err);
            }
        });
    }

    close(callback) {
        this.closed = true;
        if (this.clients) {
            for (const client of this.clients) {
                client.close();
            }
            this.clients.clear();
        }
        this.emit('close');
        if (typeof callback === 'function') {
            callback();
        }
        return this;
    }
}

module.exports = {
    WebSocket: WebSocket,
    WebSocketServer: WebSocketServer,
}

File: jsh/lib/ws/ws.go
package ws

import (
	"context"
	_ "embed"
	"fmt"
	"net/http"
	"sync"

	"github.com/dop251/goja"
	"github.com/dop251/goja_nodejs/eventloop"
	"github.com/gorilla/websocket"
	"github.com/machbase/neo-server/v8/jsh/engine"
)

//go:embed ws.js
var ws_js []byte

func Files() map[string][]byte {
	return map[string][]byte{
		"ws.js": ws_js,
	}
}

func Module(_ context.Context, rt *goja.Runtime, module *goja.Object) {
	// Export native functions
	m := module.Get("exports").(*goja.Object)
	m.Set("NewWebSocket", NewNativeWebSocket)
	m.Set("BindWebSocket", BindWebSocket)
	m.Set("CreateUpgrader", NewUpgrader)
	m.Set("IsWebSocketUpgrade", IsWebSocketUpgrade)
}

func NewNativeWebSocket(obj *goja.Object, url string, protocols []string, dispatch engine.EventDispatchFunc) *WebSocket {
	return &WebSocket{
		obj:       obj,
		url:       url,
		protocols: protocols,
		emit: func(event string, data any) {
			dispatch(obj, event, data)
		},
	}
}

type WebSocket struct {
	obj  *goja.Object
	url  string
	emit func(event string, data any)
	conn *websocket.Conn

	protocols []string
	loop      *eventloop.EventLoop

	runOnce sync.Once

	onMessage func(map[string]any)
	onClose   func(any)
	onError   func(any)
}

func NewAcceptedWebSocket(conn *websocket.Conn, loop *eventloop.EventLoop) *WebSocket {
	return &WebSocket{conn: conn, loop: loop}
}

func BindWebSocket(obj *goja.Object, raw *WebSocket, dispatch engine.EventDispatchFunc) *WebSocket {
	raw.obj = obj
	raw.emit = func(event string, data any) {
		dispatch(obj, event, data)
	}
	raw.startReadLoop()
	return raw
}

func (ws *WebSocket) Bind(obj *goja.Object, dispatch engine.EventDispatchFunc) *WebSocket {
	ws.obj = obj
	ws.emit = func(event string, data any) {
		dispatch(obj, event, data)
	}
	ws.startReadLoop()
	return ws
}

func (ws *WebSocket) OnMessage(callback func(map[string]any)) {
	ws.onMessage = callback
}

func (ws *WebSocket) OnClose(callback func(any)) {
	ws.onClose = callback
}

func (ws *WebSocket) OnError(callback func(any)) {
	ws.onError = callback
}

func (ws *WebSocket) Start() {
	ws.startReadLoop()
}

func (ws *WebSocket) RunDirect() {
	ws.run()
}

func NewUpgrader() *websocket.Upgrader {
	return &websocket.Upgrader{
		CheckOrigin: func(r *http.Request) bool {
			return true
		},
	}
}

func IsWebSocketUpgrade(r *http.Request) bool {
	return websocket.IsWebSocketUpgrade(r)
}

func (ws *WebSocket) Connect() error {
	headers := http.Header{}
	if len(ws.protocols) > 0 {
		headers.Set("Sec-WebSocket-Protocol", joinProtocols(ws.protocols))
	}
	if conn, _, err := websocket.DefaultDialer.Dial(ws.url, headers); err != nil {
		return err
	} else {
		ws.conn = conn
		ws.startReadLoop()
		return nil
	}
}

func joinProtocols(protocols []string) string {
	if len(protocols) == 0 {
		return ""
	}
	result := protocols[0]
	for i := 1; i < len(protocols); i++ {
		result += ", " + protocols[i]
	}
	return result
}

func (ws *WebSocket) Close() error {
	if ws.conn == nil {
		return nil
	}
	return ws.conn.Close()
}

func (ws *WebSocket) Send(typ int, message any) error {
	if ws.conn == nil {
		return fmt.Errorf("websocket is not open")
	}
	switch val := message.(type) {
	case string:
		return ws.conn.WriteMessage(typ, []byte(val))
	case []byte:
		return ws.conn.WriteMessage(typ, val)
	default:
		return fmt.Errorf("unsupported message type: %T", val)
	}
}

func (ws *WebSocket) Protocol() string {
	if ws.conn == nil {
		return ""
	}
	return ws.conn.Subprotocol()
}

func (ws *WebSocket) startReadLoop() {
	ws.runOnce.Do(func() {
		go ws.run()
	})
}

func (ws *WebSocket) run() {
	for {
		typ, message, err := ws.conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway, websocket.CloseNoStatusReceived) {
				payload := map[string]any{"error": err.Error()}
				if ws.onClose != nil {
					ws.dispatchCallback(func() { ws.onClose(payload) })
				} else if ws.emit != nil {
					ws.emit("close", payload)
				}
			} else {
				payload := map[string]any{"error": err.Error()}
				if ws.onError != nil {
					ws.dispatchCallback(func() { ws.onError(err) })
				} else if ws.emit != nil {
					ws.emit("error", err)
				}
				if ws.onClose != nil {
					ws.dispatchCallback(func() { ws.onClose(payload) })
				} else if ws.emit != nil {
					ws.emit("close", payload)
				}
			}
			return
		}
		data := map[string]any{
			"type": typ,
		}
		if typ == websocket.TextMessage {
			data["data"] = string(message)
		} else {
			data["data"] = message
		}
		if ws.onMessage != nil {
			ws.dispatchCallback(func() { ws.onMessage(data) })
		} else if ws.emit != nil {
			ws.emit("message", data)
		}
	}
}

func (ws *WebSocket) dispatchCallback(callback func()) {
	if ws.loop == nil {
		callback()
		return
	}
	done := make(chan struct{})
	if ok := ws.loop.RunOnLoop(func(vm *goja.Runtime) {
		defer close(done)
		callback()
	}); !ok {
		return
	}
	<-done
}
